File: src/main/scala/me/branchpanic/mods/stockpile/blockentity/TrashCanBlockEntity.scala
package me.branchpanic.mods.stockpile.blockentity

import me.branchpanic.mods.stockpile.block.StockpileProperties
import net.minecraft.block.entity.{BlockEntity, BlockEntityType}
import net.minecraft.entity.EntityType
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.inventory.SidedInventory
import net.minecraft.item.ItemStack
import net.minecraft.predicate.entity.EntityPredicates
import net.minecraft.util.Tickable
import net.minecraft.util.math.{BoundingBox, Direction}

object TrashCanBlockEntity {
  val Type: BlockEntityType[TrashCanBlockEntity] =
    BlockEntityType.Builder
      .create[TrashCanBlockEntity](() => new TrashCanBlockEntity)
      .build(null)
}

class TrashCanBlockEntity
    extends BlockEntity(TrashCanBlockEntity.Type)
    with SidedInventory
    with Tickable {
  override def getInvAvailableSlots(direction: Direction): Array[Int] = Array(1)

  override def canInsertInvStack(i: Int, itemStack: ItemStack, direction: Direction): Boolean = true

  override def canExtractInvStack(i: Int, itemStack: ItemStack, direction: Direction): Boolean =
    false

  override def getInvSize: Int = 1

  override def isInvEmpty: Boolean = true

  override def getInvStack(i: Int): ItemStack = ItemStack.EMPTY

  override def takeInvStack(i: Int, i1: Int): ItemStack = ItemStack.EMPTY

  override def removeInvStack(i: Int): ItemStack = ItemStack.EMPTY

  override def setInvStack(i: Int, itemStack: ItemStack): Unit = {}

  override def canPlayerUseInv(playerEntity: PlayerEntity): Boolean = true

  override def clear(): Unit = {}

  override def tick(): Unit = {
    if (world.isClient || !world
          .getBlockState(pos)
          .get(StockpileProperties.IsOpen)) {
      return
    }

    world
      .method_18023(EntityType.ITEM, new BoundingBox(pos.up()), EntityPredicates.VALID_ENTITY)
      .forEach(_.kill())
  }
}

File: src/main/kotlin/me/branchpanic/mods/stockpile/content/blockentity/TrashCanBlockEntity.kt
package me.branchpanic.mods.stockpile.content.blockentity

import me.branchpanic.mods.stockpile.content.block.IS_OPEN
import me.branchpanic.mods.stockpile.content.block.TrashCanBlock
import net.minecraft.block.BlockState
import net.minecraft.block.entity.BlockEntity
import net.minecraft.block.entity.BlockEntityTicker
import net.minecraft.block.entity.BlockEntityType
import net.minecraft.entity.EntityType
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.inventory.SidedInventory
import net.minecraft.item.ItemStack
import net.minecraft.predicate.entity.EntityPredicates
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.Box
import net.minecraft.util.math.Direction
import net.minecraft.world.World

class TrashCanBlockEntity(blockPos: BlockPos?, blockState: BlockState?) : BlockEntity(TYPE, blockPos, blockState), SidedInventory {
    companion object {
        val TYPE: BlockEntityType<TrashCanBlockEntity> =
            BlockEntityType.Builder.create({ blockPos: BlockPos, blockState: BlockState -> TrashCanBlockEntity(blockPos, blockState) }, TrashCanBlock).build(null)

            fun <T: BlockEntity?> tick(world: World, pos: BlockPos, state: BlockState, entity: BlockEntity?) {
                if (world.isClient) {
                    return
                }

                val currentState = world.getBlockState(pos) ?: return

                if (currentState.block != TrashCanBlock || !currentState[IS_OPEN]) {
                    return
                }

                world.getEntitiesByType(EntityType.ITEM, Box(pos.up()), EntityPredicates.VALID_ENTITY)
                    ?.forEach { e -> e.kill() }
            }
    }

    fun tick() {
        if (world == null || world?.isClient != false) {
            return
        }

        val currentState = world?.getBlockState(pos) ?: return

        if (currentState.block != TrashCanBlock || !currentState[IS_OPEN]) {
            return
        }

        world?.getEntitiesByType(EntityType.ITEM, Box(pos.up()), EntityPredicates.VALID_ENTITY)
            ?.forEach { e -> e.kill() }
    }

    override fun getStack(slot: Int): ItemStack = ItemStack.EMPTY

    override fun markDirty() = Unit

    override fun clear() = Unit

    override fun setStack(slot: Int, stack: ItemStack?) = Unit

    override fun removeStack(slot: Int): ItemStack = ItemStack.EMPTY

    override fun canPlayerUse(player: PlayerEntity?): Boolean = true

    override fun getAvailableSlots(side: Direction?): IntArray = intArrayOf(0)

    override fun size(): Int = 1

    override fun canExtract(slot: Int, stack: ItemStack?, side: Direction?): Boolean = true

    override fun removeStack(slot: Int, amount: Int): ItemStack = ItemStack.EMPTY

    override fun isEmpty(): Boolean = false

    override fun canInsert(slot: Int, stack: ItemStack?, side: Direction?): Boolean = true


}